### PR TITLE
fds: Fixed IRQ regression + IRQ timer improvements. Fixes MT8041.

### PR DESCRIPTION
**Pull request by 0kmg**

### CODE REVIEW

**Comment by 0kmg** [comment]
Regression is fixed simply by setting the clock in the driver. Timer improvements are the rest. I used Sour's FDS timer IRQ test. With just the regression fixed we get (X is fail, O is pass):
before

With improvements:
after

Unfortunately there are so many other problems with the disk system...but that's for another day.

**Comment by 0kmg** [comment]
Here's a few game fixes that I'm aware of (probably all originally regressions?):

annsmb, All Night Nippon Super Mario Brothers (Promotion Card), no longer frozen at title
almana, Almana no Kiseki, invisible status bar fixed
biomirac, Bio Miracle Bokutte Upa, status bar no longer scrolls
orudora, Chitei Tairiku Orudoora, backgrounds scroll (was unplayable)
sexyinv, Sexy Invaders, no longer frozen on first girl's picture
smb2, Super Mario Bros. 2, no longer frozen at title
timetwst, Time Twist - Rekishi no Katasumi de, no longer frozen after starting new game
wardner, Wardner no Mori, no longer frozen after starting new game
yuumaze, Yuu Maze, no longer frozen after starting new game
zeldap, Zelda no Densetsu - The Hyrule Fantasy (Prototype), no longer freezes when leaving first overworld screen

File: src/devices/bus/nes/disksys.cpp [hunk 1]
```diff
@@ -317,8 +339,9 @@ uint8_t nes_disksys_device::read_ex(offs_t offset)
 			// bit6 - End of Head (1 when disk head is on the most inner track)
 			// bit7 - Disk Data Read/Write Enable (1 when disk is readable/writable)
 			ret = m_fds_status0 | 0x80;
-			// clear the disk IRQ detect flag
-			m_fds_status0 &= ~0x01;
+			// clear the disk IRQ detect and byte transer flags
```
Comment (angelosa) [inline src/devices/bus/nes/disksys.cpp:342]: "transer"

Comment (0kmg) [reply src/devices/bus/nes/disksys.cpp:342]: thanks

**Comment by dinkc64** [comment]
0kmg, nice work!
What other problems disk sys has on mame?  I've impl'd a perfect disk system (i/o) in fbneo in early 2020, and also ported it to fceux to fix their problems.  Maybe I ran across a problem mame is having.

best regards,
- dink

**Comment by 0kmg** [comment]
Thanks, @dinkc64. I'm not actually sure what the problems are. Even as a user I haven't looked at MAME's FDS much (the above list of known bugs fixed in this PR is based off a very old list I had made before I started contributing). Now that IRQs are working again I imagine most of the remaining problems are disk I/O. Things tend to get stuck on loading. Or stuck after trying to create a new save file on Nazo no Murasame Jou, for instance. Maybe it's as simple as a few missing or incorrectly set flags in the registers? At some point I'll stop being lazy and try to figure it out. :)

**Comment by dinkc64** [comment]
Here's some things I noticed that could be improved:
instead of relying on hblank, a down-counter of 150 should be used for the transfer irq / "8th bit" status (m_fds_status0 |= 0x02;), the timer getting re-loaded'ed every read from 0x4031, or specific writes to 0x4025.
This counter is of course, in 6502 cpu cycles.

Something may help with getting writes working + something to investigate, if interested:
When saving a game, most games use this method to start writes (zelda, castlevania, super lode runner ii, etc)
0x4025: nCurrentFrame 6023  param 21   ppc e6b8 // 0x04 goes low (write-mode toggle) on param (data written to 0x4025)
0x4024: data 00                                                         // game clears shift register
0x4025: nCurrentFrame 6023  param 61   ppc e6c7 // game turns motor on (param |= 0x40)
0x4025: nCurrentFrame 6023  param e1   ppc e7a0 // game enables interrupt (param |= 0x80)
0x4024: write ignored 80, now OK to write.              // game loads dummy byte into shift register
0x4024: data 02                                                         // now we write the block ID and following data..
0x4024: data 24
0x4024: data ff

I am still not 100% sure why the dummy byte gets written to shift register at the start of writing.  From what I've read, I think / assume that we're supposed to wait until the first "8th bit" status is set from arming the first irq, when game sets param |= 0x80 on 4025.  I haven't investigate this yet, but for now simply ignoring first write to shift register after motor on works with every game.  The problem with Nazo no Murasame might be because that byte got written to the disk, and somehow the bios got confused as it verified the save (most games verify after writing).

Lastly there's the case of games which don't use normal bios routines to write files - for these to succeed, you'll need to keep track of each block length depending on the block id being accessed, including extracting the file length from the data block header to get its total block length.  I've been trying very hard to remember which games do this... as soon as it comes back to me I'll say something :)

Here's my fds I/O in-case you'd like to see how I handled the block length stuff (4024, 4025 write & 4031 read), or anything else:
https://pastebin.com/aL9xfpEi

best regards,
- dink

**Comment by 0kmg** [comment]
Ah, cool thank you. I wasn't planning on looking any time soon, but maybe with your pointers I can find time to dig in sooner than later. I'll definitely come back to your notes and code when I do.

Yeah, I knew the hblank thing was very very bad (I added the FIXME comment). I had done most of this small IRQ patch several months ago and got stuck because I was so confused as to why the transfer byte flag was triggering the interrupts in hblank (and why there was anything with hblank in FDS at all). When I revisited it a couple of days ago it made sense, in a gross hack way, as it at least somewhat approximates the disk transfer rate (about 30% over the RAM adapter's 96.4 kbs). Anyway, yeah, I was planning on putting that on a separate timer/counter. That's the only bit I knew, not having researched the rest of the FDS register details thoroughly.

As for the rest I dunno. Maybe the dummy write is something to do with the GAP? I can only speculate while skimming the docs in front of me. I know NesDev says setting bit 6 of 0x4025 is motor on/off (in addition to bit 0?!?), but the Brad Taylor doc says it's for resetting an internal CRC accumulator. Maybe the extra write is necessary for this to complete? Enri's doc in Japanese also says setting bit 6 clears a CRC register.

Oh, and I'd love to know games that go around the BIOS. That way I can ignore them and focus on getting the easy cases working :)

**Pull request merged**